### ISSUE TITLE
self hosting items

### ISSUE THREAD

**Issue by gfodor** [mozilla/hubs#1511]
- ~PWA manifest needs to be made dynamic~
- ~favicon, basic preview image need to get pushed to client~
- ~disable spoke logo on scene page~
- change APIs/perms from `hub` to `room` (greg: now that we're pushing updates automatically, I think we can punt on this -- the concern was having perma-stuck APIs on `hub`)
- ~disable twitter if no key, and media tabs if no key~
- ~add meta tags for pages in enable linking across sites~
- ~bootstrapping admin account check~
- (greg) do pass on spoke
- ~add EFS quota~
- ~restyle console~
- ~default room permissions~
- ~setup local habitat depot~
- ~always allow spoke for admins~
- ~create a on-boarding loop to hubs cloud~
- ~try dropping logos, or dropping in new default logos, for cloned repos~
- ~update email body~
- ~make all `meta` tag info (page title, taglines, etc) based upon app config in `meta` templates~
- ~add client inactivity detection and boot to both room page and admin panel~
- ~add a default `robots.txt` to minimize crawling, and allow user to override~
- ~thread through shortlink domain properly~

features to add later:
- try to get pages to be `curl`'able without hitting db
- disable room creation permission for non-admins
- pinned rooms
- landing page room
- scene-driven theme

**Comment by mqp** [comment]
This seems like a reasonable list. There are a few medium-sized ops cleanups that weigh heavily on my mind as things that need doing, which I would like to knock out to get us into a sane state before we really jump into this:

- Fix up Janus Habitat stuff. The Janus Habitat plan is in a hosed state, where using the latest version of its dependencies causes the build to barf for hard-to-understand reasons. We need to fix it. The last time we talked about this, we wanted to turn it from a giant tree of individual Habitat packages into a single package that builds its dependencies internally using a script like the one already in the plugin repository, which would eliminate this problem and eliminate all the busywork related to keeping it up to date.

- Replace Farspark CORS proxying with cors-anywhere. Robert was kvetching about this a few weeks ago because it was getting in the way of Spoke work and I felt bad that we didn't do it yet. We should clearly do it.

- Fix up logging on our boxes to work how we endorse. At the moment dd-agent makes these big spams into the log every five minutes about how it ran some checks, including things that look like errors. I don't know whether we expect to run dd-agent on our all-in-one box (maybe not?) but we should deal with this regardless.

Based on historical experience, we should also probably go make sure we're on up-to-date versions of the Habitat toolchain as best we can and fix any things that break if we try to be.

**Comment by mqp** [comment]
Here's something that I think is true but is not on the list -- we should make sure it's easy to transfer scenes and avatars between Hubs backends. Talking to Dom and Robert, it sounds like everyone is happy with the model for this being "download a GLB -> upload a GLB", but to do this we would need to expose the ability for people to upload GLBs for avatars and scenes, which we currently do not.

**Comment by mqp** [comment]
There's also some work intersecting with this which I think is a good idea, which is to separate out the Hubs client, the admin panel, and Spoke from being directly coupled with Reticulum in all the ways that they currently are (e.g. Spoke sharing the same database, Spoke and Hubs pages necessarily being served by Reticulum which mucks with them slightly) but this seems like work that should happen, if at all, after we have the big conversation about the Future Endorsed Public API Surface Of Everything.

**Comment by gfodor** [comment]
Re: above:

- I believe the fix for the janus plan is simple, just change it to bind to the last known good versions of the libraries. The reason I agreed to the shell script approach before was because I didn't understand that we could do this. Is there a reason we wouldn't do that? I see the shell script approach as going against the grain of habitat's dependency management system and insofar as we care about sharing a shell script for non-habitat users we can do that but I don't endorse us complecting that concern with writing a habitat-idiomatic plan for janus.

- I don't understand how Farspark vs cors-anywhere could be impacting/blocking anything material esp with Spoke? They both do CORS proxying through a similar API, is there some difference I'm forgetting about. The reason I thought we thought it was a good idea to get CORS proxying out of farspark was to reduce the surface area of farspark and also get that concern out of 1st party code, especially if we thought we might get off of using farspark altogether. (Which now no longer seems likely.) I don't see the upside at all but I'm not fundamentally opposed to it esp if I'm forgetting.

- I should have put metrics on the list but my general sense is there is no story there and we just won't be running any kind of dd-agent on this box. If the users want to deploy dd-agent to the machine on their own (or any other statsd compatible ingestor) they're welcome to. I don't mind us fixing it but it seems decoupled from the effort here.

- An "upload custom scene or avatar" feature seems orthogonal to this effort. It's a nice feature that will benefit all Hubs/Spoke users, including those self hosting, but I don't think we should block the efforts here on it or prioritize it over most other things here.

- I'm not sure what you're getting at with decoupling the clients further with reticulum or splitting the database. We can talk in more detail but my general sense is this is way out of scope for enabling self hosting. If you are running a fork and you also have a self hosted reticulum box you just do what we do already in production, point reticulum to the origin hosting your fork. The 'baked in' version I was referring is there as a convenience for those who aren't interested and just want to use HEAD of whatever we endorse for a given version of the server and don't care to deploy their own custom code to a s3 bucket, etc.

**Comment by gfodor** [comment]
Also re: CORS proxying I assumed our endorsed path will be to use a cloudflare worker for CORS proxying, perhaps with an option to fall back on farspark/cors-anywhere installed on the machine as a convenience. I think we should build the necessary plumbing though so it's incredibly easy for the worker to be set up properly. Eg, you should be able to go into the admin console and punch in a Cloudflare API key and domain and it should set it up for you.

**Comment by gfodor** [comment]
Also I'd like us to set an aspirational target of 9/1 as a date where we can have early adopters testing this out. I think the key to this will be smart prioritization and the main things we can't screw up are the database/account seeds (since these keys etc get burned in), the migration/upgrade process, and preventing user bricking through the tools. As long as those are sane we can dial back other stuff or push temporary solutions out before launch and improve them over time.

**Comment by mqp** [comment]
> I believe the fix for the janus plan is simple, just change it to bind to the last known good versions of the libraries. The reason I agreed to the shell script approach before was because I didn't understand that we could do this. Is there a reason we wouldn't do that? I see the shell script approach as going against the grain of habitat's dependency management system and insofar as we care about sharing a shell script for non-habitat users we can do that but I don't endorse us complecting that concern with writing a stable habitat plan for janus, which we can do by just writing the correct plan file.

To my eyes, there are three issues with our current Janus ops stuff, all largely independent:

1. We are permanently pinned to particular versions of random libraries until we fix the build breakage. Like, right now, if someone pointed out some security vulnerability in our version of OpenSSL and Habitat published a new OpenSSL package to fix it, we would have no way whatsoever to upgrade.
2. We have a large Habitat tree of packages related to Janus. This exacerbates problems like "one package changed its opinion about the version of glibc it wants and now disagrees with all the others", making it much harder to maintain than if there were a single package.
3. The versions of the libraries, build options, and configuration done through Habitat have no relationship to the versions provided by any package manager or provided by our shell script for building and installing Janus, making it harder to figure out and use whatever setup we actually endorse.

I think that "make Janus be a single Habitat package, which uses core Habitat versions of system libraries that people will already have, and runs a shell script that compiles versions of stuff we endorse for libraries that people don't already have, and make that shell script portable to run outside of Habitat too" sounds like a solution to all three problems at once, which is why I like it. Pinning the Habitat dependencies solved problem zero, which was "it was literally impossible to build the Habitat package", but it doesn't solve problems one through three.

> An "upload custom scene or avatar" feature seems orthogonal to this effort. It's a nice feature that will benefit all Hubs/Spoke users, including those self hosting, but I don't think we should block the efforts here on it or prioritize it over most other things here. Technically you can already accomplish this by uploading the GLB to a room, pinning it, and taking the URL. I think what you're suggesting here is just adding a nicer UX to enabling something like that.

I'm anticipating the following:

1. Someone spins up their Hubs-in-a-box.
2. That person notices that a scene or avatar they want from hubs.mozilla.com isn't available on their instance and wants to add it.
3. ???

This sounds like a super common thing to happen -- for example, it already happens constantly to me with prod/dev -- so I want to make sure that there is something we can type in step 3. One reasonable thing for step 3 to be would be "click a button on hubs.mozilla.com to download the GLB, and then click a button on their self-hosted avatar or scene list to upload the GLB."

> I'm not sure what you're getting at with decoupling the clients further with reticulum or splitting the database. We can talk in more detail but my general sense is this is way out of scope for enabling self hosting.

I agree with that.

**Comment by gfodor** [comment]
When I originally set up the Janus package, I did the painstaking work of writing `plan.sh` files for every single one of the missing dependencies, and their dependencies dependencies, etc. I think we can potentially solve 1 and 2 (and maybe 3) by just closing the loop and writing a cohesive script that just actually builds each of these plan files and then makes sure everything works. Is there a reason that there is suffering involved here beyond the need to manually go in and type `hab pkg build ; hab pkg upload ; hab pkg promote` on each of the dependency plans in the right order? (Which we can automate)

wrt transferring assets I'm not arguing its a bad idea or even that it won't be a common request. I'm just trying to avoid feature creep here since I want us to focus on getting a working self hosted solution in a few weeks, and the list of "must solve" problems is already quite long and challenging. so adding optionally useful things to do in front of those must solve problems seems likely to result in this taking way longer. also as i mentioned i see this as a UX/design thing not a ops task, so I see it as out of scope for this ticket/task/whatever.

**Comment by gfodor** [comment]
Also, wrt fixing up the Janus plan, you kind of tricked me past my underlying point :) The reason I was focused on "get present-day Janus to build a .hart file" as the primary goal is because doing so would unblock us on making progress towards self hosting. (I'm pretty sure we're going to need to do some non-trivial changes to the way configuration is managed, and potentially add more parameterization to the existing configs -- if so then we're blocked insofar as we cannot build a Janus package at all.)

Coming up with a more sustainable/endorsed way to package Janus going forward is a separate concern -- as long as we are using Habitat's packaging and configuration the build process we use can vary over time without any user facing impact. So I'd rather us punt on this to the degree that we can without remaining blocked on self hosting.

**Comment by misslivirose** [comment]
Closing given age and that Hubs Cloud is out - we can open new issues for future work on this front
